### PR TITLE
Provide syntax for type lambdas.

### PR DESCRIPTION
**Pull request by non**
This is a squashed version of the commits from #70.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala [hunk 1]
```diff
@@ -153,6 +153,17 @@ abstract class TreeBuilder {
   /** Create a tree representing the function type (argtpes) => restpe */
   def makeFunctionTypeTree(argtpes: List[Tree], restpe: Tree): Tree = gen.mkFunctionTypeTree(argtpes, restpe)
 
+  /** Create a type lambda tree */
+  def makeTypeLambdaTypeTree(argtpes: List[TypeDef], restpe: Tree): Tree = {
+    val name = freshTypeName("L")
+    SelectFromTypeTree(
+      CompoundTypeTree(
+        Template(
+          Select(Select(Ident("_root_"), "scala"), newTypeName("AnyRef")) :: Nil,
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala:162]: `gen.rootScalaDot(tpnme.AnyRef)`

It is inefficient to pay the lookup cost for interning standard names over an over, better to do this once in `StandardNames`.

Comment (non) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala:162]: Thanks, I'll create a patch to update it. Any other suggestions/criticisms of this patch are very welcome too!

**Comment by retronym** [comment]
Tests of the new syntax seem to be missing. Or I am missing something?

**Comment by retronym** [comment]
Overall that's the syntax we've had in mind for this. Mid term it will be cleaner and more direct to avoid desugaring to an old school type lambda, and instead carry it through in a `TypeFunction` tree, which would typecheck directly to a `PolyType`.

**Comment by retronym** [comment]
You mentioned a problem with nesting, but didn't elaborate. Can you add a commented `neg` test to show what doesn't work?

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala [hunk 2]
```diff
@@ -153,6 +153,17 @@ abstract class TreeBuilder {
...
+          ValDef(Modifiers(0), "_", TypeTree(), EmptyTree),
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala:163]: `ValDef(NoMods, nme.WILDCARD, TypeTree(), EmptyTree)`

**Comment by non** [comment]
It looks like I failed to add the files to the commit. Ugh, my bad.

I'm happy to try to produce a different tree if that's easier to work with. Honestly, I just created the old tree because I knew it "worked" and wasn't sure exactly what other tree would. I'll take a look at `TypeFunction` and see if I can just use that instead.

I'm happy to add a failing test for nesting -- it basically relates to the bug you just fixed. It was possible to do something like this:

``` scala
def foo[A[_]] = 999

foo[[x] => [y] => (x, y)]
```

which obviously doesn't really make sense, but was being accepted (since the type was not being used in `foo`).

File: src/compiler/scala/tools/nsc/ast/parser/Parsers.scala [hunk 3]
```diff
@@ -927,6 +928,7 @@ self =>
 
       /** {{{
        *  SimpleType       ::=  SimpleType TypeArgs
+       *                     |  `[' Types `]' `=>' Type
```
Comment (adriaanm) [inline src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:931]: I think this should say something like: (accepting empty `[] => Type` left as exercise to committer :-))

```
`[' VariantTypeParam {`,' VariantTypeParam} `]'] `=>' Type
```

**Comment by retronym** [comment]
Here the sort of trouble that the desugaring approach leads into:

``` scala
scala> class C { type A = String; type L[_ <: this.A] = Int; type M = L[Nothing] }
defined class C

scala> class C { type A = String; /* type M = ([_ <: this.A] => Int)[Nothing] */; }
defined class C

scala> class C { type A = String; type M = ({type L$1 [_ <: this.A] = Int})#L$1[Nothing] }
<console>:7: error: illegal cyclic reference involving type _
       class C { type A = String; type M = ({type L$1 [_ <: this.A] = Int})#L$1[Nothing] }
                                                                 ^

// Oops, we really needed:
scala> class C { type A = String; type M = ({type L$1 [_ <: C.this.A] = Int})#L$1[Nothing] }
defined class C
```

**Comment by retronym** [comment]
`TypeFunction` doesn't exist today, you will need to add it.

That will also require a trip around the neighbourhood to adjust `Traversers`, `Transformers`, pretty printers, etc.

**Comment by retronym** [comment]
In any case, this looks like an expedient first step towards a very natural syntax. Please keep us posted as to how it works out in practice when converting, say, Scalaz to use it.

Another nice-to-have would be an extension to [srewrite](https://github.com/samuelgruetter/srewrite) to automate the migration.

**Comment by adriaanm** [comment]
Agreed! Another way to experiment before introducing TypeFunction could be to use a TypeTree with a PolyType tpe.

**Comment by retronym** [comment]
Not sure if that will work, as we won't be able to construct the `PolyType` in the parser.

One hack that might work would be to use `TypeDef` tree (with `SYNTHETIC` modifier and, say, `_` as the name). You'd need to intercept that in the typechecker to discard the name and modifiers.

**Comment by adriaanm** [comment]
Ah, yes, you're right. Forgot we don't have access to types yet in the parser, which is probably for the better :-)

File: src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala [hunk 4]
```diff
@@ -153,6 +153,17 @@ abstract class TreeBuilder {
   /** Create a tree representing the function type (argtpes) => restpe */
   def makeFunctionTypeTree(argtpes: List[Tree], restpe: Tree): Tree = gen.mkFunctionTypeTree(argtpes, restpe)
 
+  /** Create a type lambda tree */
+  def makeTypeLambdaTypeTree(argtpes: List[TypeDef], restpe: Tree): Tree = {
+    val name = freshTypeName("L")
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/ast/parser/TreeBuilder.scala:158]: This should be `freshTypeName("L$")`, otherwise:

```
scala> { type L1 = Int; type T = [_] => L1 }
<console>:8: error: illegal cyclic reference involving type L1
              { type L1 = Int; type T = [_] => L1 }
                                               ^
```

**Comment by jdegoes** [comment]
Can you apply type parameters to lambdas, e.g. something like `([X] => (X, X))[Foo]` (== `(Foo, Foo)`)?

Also does application work for curried type lambdas?

**Comment by puffnfresh** [comment]
@jdegoes yes, that works.

Application does not work for curried lambdas, I would like to make that happen and I have an idea of how to do it. I'll see how feasible it is.

**Comment by jdegoes** [comment]
Awesome, looks great!

The last comment I'll make is that since the syntax mirrors the syntax for value lambdas, I think at some point, `_` should behave consistently between them (e.g. `Foo[Bar[_]]` always translates to `Foo[[X] => Bar[X]]` and not `[X] => Foo[Bar[X]]`).

Consistency between type and value lambdas will mean developers have less rules to memorize, which is why I really like the syntax that was decided on here (verbose as it may be).

**Comment by aloiscochard** [comment]
@jdegoes I think that's a bad idea due to confusion with existential, also are you sure the usage would never be ambiguous?

**Comment by jdegoes** [comment]
Existential is less common and already has `forSome` syntax (like it or not).

**Comment by aloiscochard** [comment]
It already has `_` too (like it or not).

It's really not obvious from your statements, but you propose to deprecate the usage of `_` operator for existential then?

**Comment by jdegoes** [comment]
@aloiscochard Hidden behind a flag, of course. I believe the dotc plan was to use `_` for both existential and partial application, though as many have observed, this seems ambiguous.

**Comment by nuttycom** [comment]
@aloiscochard @jdegoes I'm definitely +1 to removing _ as having any meaning as an existential and demanding explicit forSome. But that surely requires a compiler flag (could be on by default).

**Comment by propensive** [comment]
@nuttycom @aloiscochard @jdegoes You should read this thread on the Typelevel mailing list:

https://groups.google.com/d/topic/typelevel/dIyV5e-C7SI/discussion

**Comment by propensive** [comment]
TL;DR

We can infer the meaning of the `_` from the expected kind, and I'm quite strongly in favor of this approach. I spoke to @odersky about it, and he likes it too, which makes it more likely to get into mainline Scala.

**Comment by jdegoes** [comment]
If inference is totally unambiguous in all situations, I guess I don't care, but one thing I do care about is consistency between value and type lambdas, to the extent it makes sense and is possible. So looks like things are going in the right direction (from my POV).

I'm definitely looking forward to writing code like:

``` scala
// uncurried
type Flip[F[_, _]] = [A, B] => F[B, A]

Flip[Tuple2]

type Curry2[F[_, _]] = [A] => [B] => F[A, B]

// curried
type Flip'[F[_][_]] = [A] => [B] => F[B][A]

Flip'[Curry2[Tuple2]]
```

etc. :-)

**Comment by propensive** [comment]
@jdegoes I'm completely with you on consistency. It's worth sacrificing a fair amount of expressivity for consistency. @odersky thought that it's always possible to have an expected kind by which to infer the meaning of the underscore (and you would never mix existentials with partial application), but he wasn't 100% certain of this...

**Comment by non** [comment]
I must point out that even things like:

``` scala
type Flip[F[_, _]] = [A, B] => F[B, A]
```

do not currently work. I'm reluctant to try to characterize exactly what is happening, but remember that Scala doesn't currently have any notion of a type constructor that returns a type constructor. As far as I can tell, `Flip[Tuple2]` is intended to return a type constructor which can then have types applied to it.

``` scala
type FlippedTuple2[A, B]  = Tuple2[B, A]
// Flip[Tuple2] should be the same as FlippedTuple2 if I understand correctly
```

This ties back into the fact that we don't have a way to specify curried type lambdas (whose shape might be `F[_][_]` for example). Since we have @adriaanm and @retronym on the thread they may be able to say more about this (or correct any errors), but I just wanted to avoid overstating what we've gained (which is just syntax so far).

**Comment by adriaanm** [comment]
The symbol for a type constructor alias (or abstract type constructor) is a PolyType. The parameters of the PolyType are determine by the parameters of the type definition, and the result of the PolyType is a type of kind *, as specified by the RHS of the definition (for an abstract type constructor, it's a TypeBounds). There's no theoretical reason why the result couldn't be another PolyType, but that has never been the case in practice (and thus the implicit assumption is that PolyTypes are not nested).
